src/dnmd/search.template.h

```diff
@@ -21,7 +43,7 @@ static void const* SEARCH_FUNC_NAME(md_bsearch)(
     void* cxt)
 {
     assert(key != NULL && base != NULL);
-    while (count > 0)
+    while (count > 16)
```

Comment:
Why is this happening?

Reply:
This was from David's feedback (that also showed up in perf profiling) that switching to a linear search at a small number of elements ends up being better performance due to the branch predictor and the cache pre-fetching.

I'll add a comment here explaining that.

Reply:
> that switching to a linear search at a small number of elements ends up being better performance due to the branch predictor and the cache pre-fetching.

This is slightly misleading. This logic only holds if `element_size` is also reasonably small or more precisely if the X * `element_size` can fit within the cache memory. Hardcoding 16 and not incorporating the `element_size` is precisely the sort of narrow optimization that can turn degenerate later on. I think we should factor that in or assert the element size is small enough not to matter.